query_execution/QueryExecutionState.hpp

```diff
@@ -0,0 +1,294 @@
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_EXECUTION_QUERY_EXECUTION_STATE_HPP_
+#define QUICKSTEP_QUERY_EXECUTION_QUERY_EXECUTION_STATE_HPP_
+
+#include <cstddef>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+/** \addtogroup QueryExecution
+ *  @{
+ */
+
+/**
+ * @brief A class that tracks the state of the execution of a query which
+ *        includes status of operators, number of dispatched work orders etc.
+ **/
+class QueryExecutionState {
+ public:
+  /**
+   * @brief Constructor.
+   *
+   * @param num_operators Number of relational operators in the query.
+   **/
+  explicit QueryExecutionState(const std::size_t num_operators)
+      : num_operators_(num_operators), num_operators_finished_(0) {
+    done_gen_.assign(num_operators, false);
+    execution_finished_.assign(num_operators, false);
+    rebuild_required_.assign(num_operators, false);
+
+    queued_workorders_per_op_.assign(num_operators, 0);
+  }
+
+  /**
+   * @brief Get the number of operators in the query.
+   **/
+  inline const std::size_t getNumOperators() const {
+    return num_operators_;
+  }
+
+  /**
+   * @brief Get the number of operators who have finished their execution.
+   **/
+  inline const std::size_t getNumOperatorsFinished() const {
+    return num_operators_finished_;
+  }
+
+  /**
+   * @brief Check if the query has finished its execution.
+   *
+   * @return True if the query has finished its execution, false otherwise.
+   **/
+  inline bool hasQueryExecutionFinished() const {
+    return num_operators_finished_ == num_operators_;
+  }
+
+  /**
+   * @brief Set the rebuild status of the given operator that includes the
+   *        flag for whether the rebuild has been initiated and if so, the
+   *        number of pending rebuild work orders.
+   *
+   * @param operator_index The index of the given operator.
+   * @param num_rebuild_workorders The number of rebuild workorders of the given
+   *        operator.
+   * @param rebuild_initiated True if the rebuild has been initiated, false
+   *        otherwise.
+   **/
+  inline void setRebuildStatus(const std::size_t operator_index,
+                               const std::size_t num_rebuild_workorders,
+                               bool rebuild_initiated) {
+    DCHECK(operator_index < num_operators_);
```

Comment:
I prefer ``DCHECK`` because it is easier to read. With ``DCHECK_LT``, I need to refer to the signature of the function for understanding its semantics. I was forced to use ``DCHECK_GT`` and ``DCHECK_GE`` because of cpplint.